### DISCUSSION TITLE
Formatting of Inbound and Outbound Data

### DISCUSSION THREAD

**Discussion by alepping** [nebulastream/nebulastream#133] [Ideas]
# Problem
P1: Currently, formatting and parsing is done by, and sometimes even tied to (CSVSource), the source/sink, happens entirely in the runtime and is implemented non-uniformly.
P2: Currently, we assign a thread to every data source. A worker that supports dozens of sources therefore also needs to spawn dozens of threads. This can become a severe bottleneck.

# Goal
- G1: Separate formatting of inbound or outbound data from sources and sinks.
- G2: Enable compiling the formatting process of inbound/outbound data for a specific query/schema
- G3: Manage sources more scalable (not one thread per source).

# Major Discussion Threads
- [Sequence Numbers](https://github.com/nebulastream/nebulastream-public/discussions/133#discussioncomment-10093372)
- [Source sharing with different formats](https://github.com/nebulastream/nebulastream-public/discussions/133#discussioncomment-10093232)

# Ideas & Findings
## Source, Format, and Schema Combinations
Given a single worker, when separating formats from sources, we have the following possible combinations:
- source == source, format == format, schema == schema
- source == source, format == format, schema != schema
- source == source, format != format, schema == schema
- source == source, format != format, schema != schema
- source != source, format == format, schema == schema
- source != source, format == format, schema != schema
- source != source, format != format, schema == schema
- source != source, format != format, schema != schema

### Source != Source
If the sources are not equal, source sharing is not possible. However, if the format and the schema are the same, merging the queries after the sources would be possible.
### Format != Format
We should generally assume that different formats have different data layouts. Thus, if there are two different formats that have the same schema, we could merge queries after applying the formatting. Otherwise, we need to process data separately.
### Schema != Schema
If we assume that the data must arrive following the specified logical schema, then there is no chance of optimizations as long as the schema is different. A different schema means a different type of source, meaning a different formatter, meaning differently structured tuples produced by the formatter.
### Source == Source, Format == Format, Schema == Schema
This seems to be the only sane case where **source sharing** is feasible. We have two different queries that share the same source, with the same format, with the same schema on the same worker. In this case, the formatter should produce at least two copies of the formatted TupleBuffers, one for each query (given that the queries were not merged beyond the source).

### When
Given at least two queries:
- logical source == logical source
- workerID == workerID
- physical source == physical source
  -> implies:
  - source type == source type
  - source configuration == source configuration
  - format == format
  - schema == schema
  - open for sharing = True

## Source Sharing (How)
How do we decide that we can and should share sources? 

### General Sequence
1. a user marks at least one physical source as open for sharing
2. the user specifies two queries using the same logical source that contains the shareable physical source
3. the coordinator/optimizer detects that a query (q1) is already running with the specified logical source and can tag all physical sources of the other query (q2) that can be shared. The tag includes the id of q1
4. A worker W already processes q1 and receives a request to process q2 (assume the decomposed query plans are simply q1 and q2) 
5. When processing the sources during query registration, the worker checks if sources are tagged.
6. The worker then enables multiplexing the formatted buffers of the shareable sources of q1 for all sources of q1, that match the tagged sources of q2 (see [When](#when))
7. The tuples produced by the shareable sources of q1 that fulfill the [When](#when)-criteria then start producing buffers for q1 and q2 (a copy of the buffers produced for q1 is passed to q2 or similar (multiplexing)).

### Multiplexing
There are several essential aspects to consider when thinking about source sharing and multiplexing produced buffers to multiple pipelines:
- threading: currently, sources ingest data, format data, and forward produced data in a single thread
- sequence numbers: sequence numbers must reflect the order in which data arrived at the system (e.g., from a TCP source). To guarantee the correct order, we assign sequence numbers in a single thread (sequentially)
- origin ids: currently, the coordinator assigns globally unique origin ids to sources in a query in the `SourceExpansionRule`. Sources then attach the origin id to the tuple buffers they produce
- chunk numbers: formatters (e.g. a GZIP formatter) might produce more tuple buffers than they ingest. If all these tuple buffers get the same sequence id, the sequence id becomes meaningless. Hence, we need to apply chunk ids to each TupleBuffer produced from the same TupleBuffer with one specific sequence ID. The chunk id can be thought of as a nested sequence id.

### Problems
- order: if we apply formatting as an operator, that operator would become a `Task` that is executed multi-threaded, without any order guarantees. This is fatal if data might span multiple buffers (TupleBuffers can contain partial tuples or strings that are larger than a TupleBuffer) and therefore must be read and formatted in the correct order.
  - if the format operator is statufel and we assign sequence numbers at the source (when ingesting data into the system) we might be able to use these sequence numbers to read data in order 
- sequence numbers: if a shareable source of q1 already reached sequence number 1000 when q2 is deployed, q2 would start with a sequence number of 1000. We currently assume that the first sequence number is 0, which then allows us to say whether a TupleBuffer was the first (0) or whether there were prior TupleBuffers (>0). If we start with 1000, we would need to use an offset (1000) to still allow us to tell what the first TupleBuffer was.
- watermarks: TODO, where to assign watermarks?
- timestamps: TODO, where to assign timestamps

# Solution
Todo (just a very rough sketch below):
## S1
We create FormatOperator from the SourceDescriptor at some point during the lowering. An example would be the FormatOperatorCSV.\
Sources read raw data via a specific protocol or from a file and write the raw bytes into a TupleBuffer.\
The source forwards the raw data to the FormatOperator that then applies a specific formatting.

## S2
Using runtime information, we can implement the FormatOperator in Nautilus to a certain degree. The Nautilus code is then compiled to machine code.

## S3
Potentially, we can spawn one thread that alternates between sources and reads data from the different interfaces and pushes Tasks to the queue.

**Comment by keyseven123** [comment]
> Format != Format
> We should generally assume that different formats have different data layouts. Thus, if there are two different formats that have the same schema, we could merge queries after applying the formatting. Otherwise, we need to process data separately.

Why can we not share the physical source here, if the source is decoupled from the format itself? 
We just have to take care of the sequence numbering. Anything else, e.g., formatting, setting the origin id, can happen out-of-order in a multi-threaded fashion.

**Reply by alepping** [reply]
@keyseven123 
> Why can we not share the physical source here, if the source is decoupled from the format itself?

If we allowed for sharing with different formats, we would for example allow a single TCP source to process GZip data and CSV data.\
Physically, the TCP source could of course do that, since it only processes bytes and does not care about the underlying format.\
However, how would we then decide whether data that arrived at the TCP source is GZip data or CSV data? Packages containing binary GZip data and CSV data might even be mixed in a single TupleBuffer.

In general, logically tying formats to sources makes a lot of sense in my opinion. Hence why I defined the source/sink descriptor in #64 as having:
- source type
- format
- (optional) configuration
- (optional) metadata

Physically, we should still perform the formatting in a separate step.

**Reply by PhilippGrulich** [reply]
Hey maybe I have a interesting practical example. At work we receive data in a single raw stream where each entry is a string. However, some stings are json, some are just log lines. So the input format is very naive but the actual data is extracted in the processing pipeline.
I think this would also work with your proposal.

**Reply by adrianmichalke** [reply]
That's an interesting example. @alepping does telegraf support such cases? This would require logic to decide ad hoc which formatting to use. I wonder if we can come up with a generalized logic for this. 

A log format itself is also very interesting. I guess there is no general 'log format' that we can support. Would that be a case where external format plugins are needed?

**Reply by ls-1801** [reply]
I think this would require a more general parse expression or UDFs.

```cpp
Query::from(log)
  .filter(isLogWithJson(Attribute("data")))
  .map(unpackJson(Attribute("id").as(UINT64), Attribute("name").as(TEXT)))
```

**Reply by alepping** [reply]
@PhilippGrulich 
> some stings are json, some are just log lines

Interesting! Essentially, you have a byte stream, but some bytes in the byte stream need to be interpreted (formatted) differently than other bytes in the byte stream? Or, given that you say 'strings', is it really a stream of strings and the size of the strings, and their format proceeds the raw bytes of the strings?

**Reply by alepping** [reply]
@adrianmichalke 
> This would require logic to decide ad hoc which formatting to use.

I agree. I thought that there are few use cases for that, but potentially, given @PhilippGrulich 's example, this is not the case.
I still see it an optimization down the line though, since it requires a more general parser, as @ls-1801 mentioned (as far as I can see). I would first like to keep it simple and tie a physical source to one format.

**Reply by PhilippGrulich** [reply]
> I think this would require a more general parse expression or UDFs.
> 
> 
> 
> ```cpp
> 
> Query::from(log)
> 
>   .filter(isLogWithJson(Attribute("data")))
> 
>   .map(unpackJson(Attribute("id").as(UINT64), Attribute("name").as(TEXT)))
> 
> ```
> 
> 

Yes basically some systems have something like PARSE_JSON, which receives a string as input and if it is not a valid json the result is NULL.

**Comment by keyseven123** [comment]
> sequence numbers: sequence numbers must reflect the order in which data arrived at the system (e.g., from a TCP source). To guarantee the correct order, we assign sequence numbers in a single thread (sequentially)

Not only this, but every stateful operator must know what the first sequence number is. Currently, we achieve this by assuming that the first sequence number is always 1. 
- It is not clear in the solution draft, who and where the sequence number is set. 
- Also, do we still stick to starting the sequence number for every query from 1? 

One idea is to have the following flow:
```mermaid
flowchart TD
    A(TCPPhysicalSource) -->|1. Acquire TB| BM(BufferManager)
    A -->|2. SeqNumber = getSeqForQuery| A
    A -->|3. RawBytesTB + SeqNumber| FMT(Formatter)
    FMT --> |4. Passing TB| Scan(ScanOperator) 
```
I think that the source takes some raw bytes and encapsulates them into a tuple buffer together with a sequence number for every format/query that expects one, e.g. the steps `2.` and `3.` are a loop over all successor pipelines.
```cpp

for (auto formatPerQuery : allSuccessorPipelines) {
    emitRawBytes(TupleBuffer& rawBytes, SequenceNumber sequenceNumber);
}
```

This way, every successor pipeline / query gets a sequence number starting from `1` but the formatting can be done in a multithreaded way.

**Reply by alepping** [reply]
@keyseven123 
>  It is not clear in the solution draft, who and where the sequence number is set.

You are right. I wrote it between the lines but not at all clearly.
Your mermaid diagram depicts essentially what I envision, only that there is something like a `DataSourceHandle` that is initiated using a specific source, e.g. the TCPPhysicalSource, and that then handles acquiring the TB and getting a sequence number:

```mermaid
flowchart TD
    A["DataSourceHandle< TCPSource >.start()"] --> |1. Acquire TB| BM(BufferManager)
    A -->|2. SeqNumber = getSeqForQuery| A
    A -->|3. RawBytesTB + SeqNumber| FMT(Formatter)
    FMT --> |4. Passing TB| Scan(ScanOperator) 
```

**Reply by alepping** [reply]
> This way, every successor pipeline / query gets a sequence number starting from 1 but the formatting can be done in a multithreaded way.

I still see the problem that the formatter requires that TupleBuffers arrive in the correct order, otherwise it cannot format data correctly, e.g. if variable sized data spans over multiple TupleBuffers.

**Reply by ls-1801** [reply]
@keyseven123 wouldn't this mean we do the formatting multiple times?

**Reply by keyseven123** [reply]
> > This way, every successor pipeline / query gets a sequence number starting from 1 but the formatting can be done in a multithreaded way.
> 
> I still see the problem that the formatter requires that TupleBuffers arrive in the correct order, otherwise it cannot format data correctly, e.g. if variable sized data spans over multiple TupleBuffers.

If we support arbitrary large variable sized data, then the full processing must happen sequentially. Maybe, we can at least parallelize it over multiple physical streams, as each physical stream sets the sequence number.

**Reply by keyseven123** [reply]
> @keyseven123 wouldn't this mean we do the formatting multiple times?

No, it should be formatted once, either as a standalone module or in the scan operator.

**Reply by ls-1801** [reply]
But what does this `emitRawBytes` function do? Does it submit a new Task into the Pipeline?

**Reply by alepping** [reply]
@ls-1801 
I understand it the same way, emitRawbytes essentially creates a Task with the format pipeline and a TupleBuffer containing raw bytes. However, a Task could mean asynchronous processing, which is problematic.

**Reply by ls-1801** [reply]
@alepping but if we emit the raw buffer for every successor aren't we doing the parsing multiple times?
